pkg/cloud/request.go

```diff
@@ -0,0 +1,196 @@
+	stop := make(chan struct{})
+	delayedCtx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		select {
+		case <-stop:
+			return
+		case <-parentCtx.Done():
+			time.Sleep(delay)
+			cancel()
+		}
+	}()
+
+	return &delayedCancelContext{
+		Context: delayedCtx,
+		cancel:  cancel,
+		stop:    stop,
+	}
+}
+
+type BackoffResource struct {
+	Type string
+	Name string
+}
+
+type BackoffSettings struct {
+	Enabled             bool          `cfg:"enabled"`
+	Blocking            bool          `cfg:"blocking"`
+	CancelDelay         time.Duration `cfg:"cancel_delay"`
+	InitialInterval     time.Duration `cfg:"initial_interval"`
+	RandomizationFactor float64       `cfg:"randomization_factor"`
+	Multiplier          float64       `cfg:"multiplier"`
+	MaxInterval         time.Duration `cfg:"max_interval"`
+	MaxElapsedTime      time.Duration `cfg:"max_elapsed_time"`
+}
+
+type AwsRequestFunction func(ctx context.Context) (interface{}, error)
+
+type RequestExecutor interface {
+	Execute(ctx context.Context, f AwsRequestFunction) (interface{}, error)
+}
+
+type DefaultExecutor struct {
+}
+
+func (e DefaultExecutor) Execute(ctx context.Context, f AwsRequestFunction) (interface{}, error) {
+	return f(ctx)
+}
+
+type BackoffExecutor struct {
+	logger   mon.Logger
+	metric   mon.MetricWriter
+	res      *BackoffResource
+	settings *BackoffSettings
+}
+
+func NewBackoffExecutor(logger mon.Logger, res *BackoffResource, settings *BackoffSettings) *BackoffExecutor {
+	defaults := getBackoffExecutorDefaultQueueMetrics(res)
+	metric := mon.NewMetricDaemonWriter(defaults...)
+
+	return &BackoffExecutor{
+		logger:   logger,
+		metric:   metric,
+		res:      res,
+		settings: settings,
+	}
+}
+
+func (e *BackoffExecutor) Execute(ctx context.Context, f AwsRequestFunction) (interface{}, error) {
+	logger := e.logger.WithContext(ctx)
+
+	cancelCtx, _ := context.WithCancel(ctx)
+
+	delayedCtx := WithDelayedCancelContext(ctx, e.settings.CancelDelay)
+	defer delayedCtx.Stop()
+
+	if !e.settings.Enabled {
+		return f(delayedCtx)
+	}
+
+	var out interface{}
+	var err error
+
+	backoffConfig := backoff.NewExponentialBackOff()
+
+	if e.settings.InitialInterval > 0 {
+		backoffConfig.InitialInterval = e.settings.InitialInterval
+	}
+
+	if e.settings.RandomizationFactor > 0 {
+		backoffConfig.RandomizationFactor = e.settings.RandomizationFactor
+	}
+
+	if e.settings.Multiplier > 0 {
+		backoffConfig.Multiplier = e.settings.Multiplier
+	}
+
+	if e.settings.MaxInterval > 0 {
+		backoffConfig.MaxInterval = e.settings.MaxInterval
+	}
+
+	if e.settings.MaxElapsedTime > 0 {
+		backoffConfig.MaxElapsedTime = e.settings.MaxElapsedTime
+	}
+
+	if e.settings.Blocking {
+		backoffConfig.MaxElapsedTime = 0
+	}
+
+	backoffCtx := backoff.WithContext(backoffConfig, cancelCtx)
+
+	notify := func(err error, duration time.Duration) {
+		logger.Warnf("retrying aws service %s %s after error: %s", e.res.Type, e.res.Name, err.Error())
```

Comment:
Will this trigger an alarm?

Reply:
plan is to adjust the warning alarms that they are not triggering every single time